[DDW-1152] Implement acceptance tests for custom number, date and time formats

daniloprates opened the pull request:
This PR implements acceptance tests for custom number, date and time formats.

## Todos

 - [x] Initial profile setup is working as expected (user needs to select number/date/time format)
 - [x] Changing number/date/time format on the "Settings > General" screen works as expected
 - [x] All dates are displayed in the correct format based on user preference
 - [x] All numbers are displayed in the correct format based on user preference (including wallet sidebar wallet balance display)
 - [x] Users can create transactions regardless of their number format preference (including testing amounts bigger than 1 million ada)

## Screenshots

![image](https://user-images.githubusercontent.com/1504716/75683914-b6bda680-5c76-11ea-8b09-1bf2aaa098f4.png)


---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1583158551081000)
- [ ] Test

---

## Review Checklist

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes and shows some screenshots or animated GIFs of important UI changes
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

yakovkaravelov commented on tests/settings/e2e/steps/custom-formats.js:
```diff
@@ -0,0 +1,90 @@
+When(/^the "([^"]*)" wallet has received the transaction amount$/, async function(walletName) {
+  await this.client.waitUntil(async () => {
+    const currentAmount = await getValueFromSelector.call(this, walletName, 'number')
+    console.log('currentAmount', currentAmount);
```
@daniloprates Please remove this `console.log`

daniloprates replied:
@yakovkaravelov ✅

yakovkaravelov commented on tests/settings/e2e/features/custom-formats.feature:
```diff
@@ -0,0 +1,157 @@
+@e2e  @watch
```
@daniloprates Please remove `@watch` tag.

daniloprates replied:
@yakovkaravelov ✅

nikolaglumac commented:
@tomothespian please review this one 🙏

daniloprates commented:
> @daniloprates Great work, just left minor comments. 👍

Thanks @yakovkaravelov. Didn't realize `lint` doesn't catch console.logs inside tests.

Fixed ✅

tomothespian commented on tests/settings/e2e/steps/custom-formats.js:
```diff
@@ -0,0 +1,89 @@
+        daedalus.stores.profile._updateUserLocalSetting({ param: `${param}Format`, value }))
+    ).then(done);
+  }, chosenFormats);
+  await timeout(1500);
```
@daniloprates is `timeout()` necessary here?

daniloprates replied:
✅

tomothespian commented on tests/settings/e2e/steps/custom-formats.js:
```diff
@@ -0,0 +1,89 @@
+When(/^the "([^"]*)" wallet has received the transaction amount$/, async function(walletName) {
+  await this.client.waitUntil(async () => {
+    const currentAmount = await getValueFromSelector.call(this, walletName, 'number')
+    return currentAmount !== '0 ADA';
```
@daniloprates maybe here would be better to check that displaied `currenAmount` is equal to transaction amount from the store

daniloprates replied:
✅

tomothespian commented on tests/settings/e2e/steps/helpers.js:
```diff
@@ -89,3 +126,69 @@ export const termsOfUseHelpers = {
+
+export const getValueFromSelector = async function  (screenElement: string, expectedParam: string) {
+  const selector = screenElementSelectors[screenElement][expectedParam];
+  // await this.client.waitForVisible(selector);
```
@daniloprates please remove the comment

daniloprates replied:
✅

tomothespian commented on source/renderer/app/components/wallet/transactions/Transaction.js:
```diff
@@ -379,7 +379,7 @@ export default class Transaction extends Component<Props, State> {
                 <div className={styles.type}>
                   {intl.formatMessage(messages.type, { currency })},{' '}
-                  {moment(data.date).format(currentTimeFormat)}
+                  <em>{moment(data.date).format(currentTimeFormat)}</em>
```
@daniloprates maybe we shouldn't change code that we want to test just to fit tests. Is this some requirement not related to tests?
If you need that for test and if is impossible to get that in scenario rather use React.Fragment than <em> which will transform text!

daniloprates replied:
✅

tomothespian commented:
@daniloprates I got this few times. Sometimes pass and sometimes not.
<img width="943" alt="Screen Shot 2020-03-03 at 13 21 25" src="https://user-images.githubusercontent.com/18653950/75776011-6b2cfa80-5d53-11ea-8338-2e0228f1762d.png">

daniloprates commented:
> @daniloprates Amazing work. 🚀
> I left just a few comments. Can you please check?
> Also, when I run all tests together I get one failed test but not related to this task. I will run again and let you know!

@tomothespian thanks for the review and requests. I've fixed and adjusted them.

tomothespian commented:
@daniloprates Nice improvements, thanks.
But I still get this error:
<img width="1342" alt="Screen Shot 2020-03-03 at 15 23 29" src="https://user-images.githubusercontent.com/18653950/75784857-1f824d00-5d63-11ea-978c-ab150098d79b.png">

daniloprates commented:
> @daniloprates Nice improvements, thanks.
> But I still get this error:
> <img alt="Screen Shot 2020-03-03 at 15 23 29" width="1342" src="https://user-images.githubusercontent.com/18653950/75784857-1f824d00-5d63-11ea-978c-ab150098d79b.png">

@tomothespian I could never reproduce it 😕 but made some adjustments to try avoiding it.

tomothespian commented:
@daniloprates hmm weird. I will try one more time.

tomothespian commented:
@daniloprates it is really weird. When I run just run your new tests with **@watch** flag all tests pass but when I run all tests together then always fails on the step I reported above :(

daniloprates commented:
> @daniloprates it is really weird. When I just run your new tests with **@watch** flag all tests pass but when I run all tests together then always fails on the step I reported above :(

Fixed ✅

tomothespian commented:
@daniloprates thanks. Now your tests works but when I run all tests together always fails on same scenario:
<img width="880" alt="Screen Shot 2020-03-04 at 10 41 05" src="https://user-images.githubusercontent.com/18653950/75866062-c66af580-5e04-11ea-865f-7ef0a7e334f3.png">

Can you please check?

daniloprates commented:
> @daniloprates thanks. Now your tests works but when I run all tests together always fails on same scenario:
> Can you please check?

@tomothespian fixed ✅ I'll surely go into more detail on my current task `DDW-1213 Optimize E2E tests`.

nikolaglumac commented:
@tomothespian can you please re-test this one? 🙏

tomothespian commented:
@daniloprates same thing happened again for step: `Successfully transferring funds from "Daedalus Balance" wallet to "Daedalus Rewards" wallet`

daniloprates commented:
> @daniloprates same thing happened again for step: `Successfully transferring funds from "Daedalus Balance" wallet to "Daedalus Rewards" wallet`

@tomothespian this test has failed numerous times on `v2-integration` as well, for me. I'm working on "DDW-1213 Optimize E2E tests" and will tackle this and other ones that are unrelated to `DDW-1868`.

nikolaglumac commented:
@daniloprates please merge latest v2-integration - it looks like tests are more stable with the latest version of the cardano-wallet and Jormungandr which is available in v2-integration

daniloprates commented:
> @daniloprates please merge latest v2-integration - it looks like tests are more stable with the latest version of the cardano-wallet and Jormungandr which is available in v2-integration

@nikolaglumac the tests I'm fixing now here are not related to this PR. I'd rather merge this one and fix them on "DDW-1213 Optimize E2E tests" if it's ok.
